lib/utils/downloadTemplateFromRepo.js

```diff
@@ -0,0 +1,196 @@
+      ' https://github.com/serverless/serverless',
+    ].join('');
+    throw new ServerlessError(errorMessage);
+  }
+
+  let pathToDirectory = '';
+  for (let i = 5; i <= (parts.length - 1); i++) {
+    pathToDirectory = path.join(pathToDirectory, parts[i]);
+  }
+
+  const downloadUrl = [
+    'https://bitbucket.org/',
+    owner,
+    '/',
+    repo,
+    '/get/',
+    branch,
+    '.zip',
+  ].join('');
+
+  return {
+    owner,
+    repo,
+    branch,
+    downloadUrl,
+    isSubdirectory,
+    pathToDirectory,
+  };
+}
+
+/**
+ * Parse URL and call the appropriate adaptor
+ *
+ * @param {string} inputUrl
+ * @throws {ServerlessError}
+ * @returns {Object}
+ */
+function parseRepoURL(inputUrl) {
+  if (!inputUrl) {
+    throw new ServerlessError('URL is required');
+  }
+
+  const url = URL.parse(inputUrl.replace(/\/$/, ''));
+
+  // check if url parameter is a valid url
+  if (!url.host) {
+    throw new ServerlessError('The URL you passed is not a valid URL');
+  }
+
+  switch (url.hostname) {
+    case 'github.com': {
+      return parseGitHubURL(url);
+    }
+    case 'bitbucket.org': {
+      return parseBitBucketURL(url);
+    }
+    default: {
+      const msg = 'The URL you passed is not one of the valid providers: "GitHub", "BitBucket".';
+      throw new ServerlessError(msg);
+    }
+  }
+}
+
+/**
+ * @param {string} inputUrl
+ * @param {string} [templateName]
+ * @param {string} [path]
+ * @returns {Promise}
+ */
+function downloadTemplateFromRepo(inputUrl, templateName, downloadPath) {
+  const repoInformation = parseRepoURL(inputUrl);
+
+  let serviceName;
+  let dirName;
+  let downloadServicePath;
+
+  if (repoInformation.isSubdirectory) {
+    const folderName = repoInformation.pathToDirectory.split('/').splice(-1)[0];
+    serviceName = folderName;
+    dirName = downloadPath || templateName || folderName;
+    downloadServicePath = path.join(os.tmpdir(), repoInformation.repo);
+  } else {
+    serviceName = repoInformation.repo;
+    dirName = downloadPath || templateName || repoInformation.repo;
+    downloadServicePath = path.join(process.cwd(), dirName);
+  }
+
+  const servicePath = path.join(process.cwd(), dirName);
+  const renamed = dirName !== repoInformation.repo;
+
+  if (dirExistsSync(path.join(process.cwd(), dirName))) {
+    const errorMessage = `A folder named "${dirName}" already exists.`;
+    throw new ServerlessError(errorMessage);
+  }
+
+  console.log(`Serverless: ${chalk.yellow(`Downloading and installing "${serviceName}"...`)}`);
+
+  // download service
+  return download(
+    repoInformation.downloadUrl,
+    downloadServicePath,
+    { timeout: 30000, extract: true, strip: 1, mode: '755' }
+  ).then(() => {
+    // if it's a directory inside of git
+    if (repoInformation.isSubdirectory) {
+      const directory = path.join(downloadServicePath, repoInformation.pathToDirectory);
+      copyDirContentsSync(directory, servicePath);
```

Comment:
Usually, we try to get rid of `sync` calls in favor of `async` ones but I don't think that's an issue here. It's not an integral part of framework invoked very frequently.